packages/celotool/src/e2e-tests/validator_order_tests.ts

```diff
@@ -0,0 +1,109 @@
+const VALIDATOR_BLOCK_COUNT_ARRAY = Array(VALIDATORS).fill(VALIDATOR_BLOCKS)
+
+describe('governance tests', () => {
+  const gethConfig: GethTestConfig = {
+    migrateTo: 13,
+    instances: [
+      { name: 'validator0', validating: true, syncmode: 'full', port: 30303, rpcport: 8545 },
+      { name: 'validator1', validating: true, syncmode: 'full', port: 30305, rpcport: 8547 },
+      { name: 'validator2', validating: true, syncmode: 'full', port: 30307, rpcport: 8549 },
+      { name: 'validator3', validating: true, syncmode: 'full', port: 30309, rpcport: 8551 },
+      { name: 'validator4', validating: true, syncmode: 'full', port: 30311, rpcport: 8553 },
+    ],
+  }
+
+  const context: any = getContext(gethConfig)
+  let contractKit: ContractKit
+
+  before(async function(this: any) {
+    this.timeout(0)
+    await context.hooks.before()
+  })
+
+  after(context.hooks.after)
+
+  const restart = async () => {
+    await context.hooks.restart()
+    contractKit = newKit('http://localhost:8545')
+    contractKit.defaultAccount = validatorAddress
+
+    // TODO(mcortesi): magic sleep. without it unlockAccount sometimes fails
+    await sleep(2)
+  }
+
+  describe('Validator ordering', () => {
+    before(async function() {
+      this.timeout(0)
+      await restart()
+    })
+
+    it('properly orders validators randomly', async function(this: any) {
+      this.timeout(100000)
+
+      const latestBlockNumber = (await contractKit.web3.eth.getBlock('latest')).number
+      const epochRemainder = latestBlockNumber % EPOCH
+      const nextEpoch = latestBlockNumber + (epochRemainder === 0 ? 0 : EPOCH - epochRemainder) + 1
+
+      // wait for enough blocks
+      while ((await contractKit.web3.eth.getBlock('latest')).number < nextEpoch + BLOCK_COUNT) {
+        await sleep(2)
+      }
+
+      // fetch the validator for each block
+      const blocks = await Promise.all(
+        BLOCK_COUNT_ARRAY.map(async (num) => contractKit.web3.eth.getBlock(num + nextEpoch))
+      )
+      const validators = blocks.map((block) => block.miner)
+
+      // ensure each validator has an equal number of blocks
+      const validatorCounts = _.countBy(validators)
+      assert.deepEqual(Object.values(validatorCounts), VALIDATOR_BLOCK_COUNT_ARRAY)
+
+      let firstOrdering
+      let uniqueOrdering = false
+      for (let i = 0; i < EPOCHS_TO_WAIT; i++) {
+        const epochBlocks = validators.slice(i * EPOCH, (i + 1) * EPOCH)
+        const ordering = epochBlocks.slice(0, EPOCH / VALIDATORS)
+        if (i === 0) {
+          firstOrdering = ordering
+        }
+
+        // ensure within an epoch, ordering is consistent
+        ordering.forEach((validator, index) => {
+          assert.equal(validator, epochBlocks[VALIDATORS + index])
+        })
+
+        // ensure there is a unique ordering in different epochs
+        if (i !== 0) {
+          if (!_.isEqual(firstOrdering, ordering)) {
+            uniqueOrdering = true
```

Comment:
This does have a `1/5! = 0.8%` chance of failing. I would say thats not quite low enough considering we might reasonably run this test more than 100 times per month. Increasing the number of validators, for example to 10, would pretty quickly reduce that risk.